whitepapers/0013_ccq.md

```diff
@@ -0,0 +1,350 @@
+- `ccqP2pBootstrap` - bootstrap peers for the CCQ P2P channel. No default (but auto generated in tilt).
+- `ccqAllowedPeers` - comma separated list of P2P peer IDs that are allowed to submit query requests.
+
+### No Query Persistence in the Guardian
+
+In order to reduce the storage burden on the guardian node, full responses are not persisted in the guardian. However, to facilitate de-duplication and authorization, some cross-chain query information may be committed to Wormchain.
+
+## CCQ REST Server
+
+Some integrators might choose to communicate directly with the guardians by encoding their own query requests and publishing them to the gossip network. To do so would require their P2P peer ID to be configured on the guardians. Additionally they would need to have a P2P library, which limits their choices (For instance, there does not currently exist a typescript library for P2P). Additionally, the integrator would need to collect and consolidate the responses.
+
+For these and other reasons, a separate REST server is being provided to accept off-chain query requests, submit them to the gossip network, correlate the responses, and return them to the caller.
+
+The CCQ server will be configured to publish and listen on the CCQ P2P network. It's P2P peer ID will be registered with the guardians. The server will join both `ccq_req` and `ccq_resp` but only subscribe to `ccq_resp`.
+
+The server will be configured with a list of supported users. Each user will have an API Key, as well as a list of what queries they are allowed to perform.
+The query data per user will look something like this:
+
+```
+{
+    "userName": "Test User",
+    "apiKey": "432185ac-fbb7-4883-b68d-888e7bd749a9",
+    "allowUnsigned": true,
+    "allowedCalls": [
+    {
+        "ethCall": {
+        "note:": "Name of WETH on Goerli",
+        "chain": 2,
+        "contractAddress": "B4FBF271143F4FBf7B91A5ded31805e42b2208d6",
+        "call": "0x06fdde03"
+        }
+    }
+    ]
+}
+```
+
+When a user submits a query, they will need to set the `X-API-Key` header in the request.
+
+The server will only accept requests from configured users. Those requests may only be for the RPC calls configured in the `allowedCalls` list.
+
+Each allowed call is specified by:
+
+- Target wormhole chain ID
+- Target contract address on that chain
+- The first four bytes of the hash of signature of the method to be called.
+
+All configured users may submit queries that they sign with their own key. In addition to signed requests, if the `allowUnsigned` flag is set to `true`, the user may submit unsigned requests and the server will sign them using a pre-configured key. Note that all keys must be in the guardian allow list.
+
+## Typescript Library
+
+To assist integrators, there is a typescript library in `sdk/js-query` that provides methods for building queries. Additionally, there are some usage examples in `sdk/js-query/src/query/ethCall.test.ts`.
```

Comment:
Will these be deployed to NPM separately?

Reply:
Yes, these will be published to npm separately to reduce the package size. Once published, we should come back and put the link in this section.